libdnf/dnf-repo.c

```diff
@@ -737,6 +753,101 @@ dnf_repo_set_keyfile(DnfRepo *repo, GKeyFile *keyfile)
 }
 
 /**
+ * dnf_repo_set_metadata_expire:
+ * @repo: a #DnfRepo instance.
+ * @metadata_expire: the expected expiry time for metadata
+ *
+ * Sets the metadata_expire time, which is default to be 0 days
+ **/
+void
+dnf_repo_set_metadata_expire(DnfRepo *repo, guint metadata_expire)
+{
+    DnfRepoPrivate *priv = GET_PRIVATE(repo);
+    priv->metadata_expire = metadata_expire;
+}
+
+/**
+ *  dnf_repo_parse_time_from_str
+ *  @expression: a expression to be parsed
+ *  @error: error item
+ *
+ *  parser to parse string into an integer value of seconds, or a human
+ *  readable variation specifying days, hours, minutes or seconds
+ *  until something happens. Works like :class:`BytesOption`.  Note
+ *  that due to historical president -1 means "never", so this accepts
+ *  that and allows the word never, too.
+ *
+ *  Valid inputs: 100, 1.5m, 90s, 1.2d, 1d, 0xF, 0.1, -1, never.
+ *  Invalid inputs: -10, -0.1, 45.6Z, 1d6h, 1day, 1y.
+
+ *  Return value will always be an integer
+ **/
+
+static guint dnf_repo_parse_time_from_str(const gchar *expression, GError **error)
+{
+    gchar last_char = expression[ strlen(expression) - 1 ];
+    gint multiplier;
+    guint result;
+    gdouble parsed_time;
+    gchar *endptr = NULL;
+
+    if (strlen(expression) < 1) {
+        g_set_error_literal(error,
+                            DNF_ERROR,
+                            DNF_ERROR_FILE_INVALID,
+                            "no value specified");
+        return 0;
+    }
+
+    if (g_strcmp0(expression, "-1") == 0 || g_strcmp0(expression,"never") == 0)
+        return G_MAXUINT;
+
+    /* check if the input ends with h, m ,d ,s as units */
+    if (g_ascii_isalpha(last_char)) {
+        if(last_char == 'h')
+            multiplier = 60 * 60;
+        else if(last_char == 's')
+            multiplier = 1;
+        else if(last_char == 'm')
+            multiplier = 60;
+        else if(last_char == 'd')
+            multiplier = 60 * 60 * 24;
+        else {
+            g_set_error(error, DNF_ERROR, DNF_ERROR_FILE_INVALID,
+                        "unknown unit %c", last_char);
+            return 0;
+            }
+    }
+    else
+        multiplier = 1;
+
+    /* convert expression into a double*/
+    parsed_time = g_ascii_strtod(expression, &endptr);
+
+    /* failed to parse */
+    if (expression == endptr) {
+        g_set_error(error, DNF_ERROR, DNF_ERROR_INTERNAL_ERROR,
+                    "failed to parse time: %s", expression);
+        return 0;
+    }
+
+    /* too large */
+    if (parsed_time > G_MAXDOUBLE) {
+        g_set_error(error, DNF_ERROR, DNF_ERROR_INTERNAL_ERROR,
+                    "time too large");
+        return 0;
+    }
+    /* time can not be below zero */
+    if (parsed_time < 0) {
+        g_set_error(error, DNF_ERROR, DNF_ERROR_INTERNAL_ERROR,
+                    "seconds value must not be negative %s",expression );
+        return 0;
+    }
+
+    result = (guint) (parsed_time * multiplier);
```

Comment:
Minor nit, but you could drop the `result` variable and simply do `return (guint)(parsed_time * multiplier)`.